src/Features/Core/Portable/PullMemberUp/MembersPuller.cs

```diff
@@ -0,0 +1,408 @@
+            }
+        }
+
+        private async Task<Solution> PullMembersIntoClassAsync(
+            Document document,
+            PullMembersUpAnalysisResult result,
+            CancellationToken cancellationToken)
+        {
+            var solution = document.Project.Solution;
+            var solutionEditor = new SolutionEditor(solution);
+            var codeGenerationService = document.Project.LanguageServices.GetRequiredService<ICodeGenerationService>();
+            var destinationSyntaxNode = await codeGenerationService.FindMostRelevantNameSpaceOrTypeDeclarationAsync(
+                document.Project.Solution, result.Destination, default, cancellationToken).ConfigureAwait(false);
+            var (editorMap, syntaxMap) = await InitializeEditorMapsAndSyntaxMapAsync(result, solution, solutionEditor, destinationSyntaxNode, cancellationToken).ConfigureAwait(false);
+            // Add members to destination
+            var pullUpMembersSymbols = result.MemberAnalysisResults.SelectAsArray(
+                memberResult =>
+                {
+                    if (memberResult.MakeDeclarationAtDestinationAbstract)
+                    {
+                        // Change the member to abstract if user choose to make them abstract
+                        return GetAbstractMemberSymbol(memberResult.Member);
+                    }
+                    else
+                    {
+                        return memberResult.Member;
+                    }
+                });
+            var options = new CodeGenerationOptions(reuseSyntax: true, generateMethodBodies: false);
+            var newDestination = codeGenerationService.AddMembers(destinationSyntaxNode, pullUpMembersSymbols, options: options);
+
+            // Remove some original members since we are pulling members into class.
+            // Note: If user chooses to make the member abstract, then the original member won't be touched,
+            // It will just pull abstract declaration up to destination.
+            // But if the member is abstract itself, it will still be removed.
+            foreach (var analysisResult in result.MemberAnalysisResults)
+            {
+                if (!analysisResult.MakeDeclarationAtDestinationAbstract)
+                {
+                    foreach (var syntax in syntaxMap[analysisResult.Member])
+                    {
+                        var originalMemberEditor = editorMap[syntax.SyntaxTree];
+                        originalMemberEditor.RemoveNode(originalMemberEditor.Generator.GetDeclaration(syntax));
+                    }
+                }
+            }
+
+            // Change the destination to abstract class if needed.
+            var destinationEditor = editorMap[destinationSyntaxNode.SyntaxTree];
+            if (!result.Destination.IsAbstract &&
+                result.MemberAnalysisResults.Any(analysis => analysis.Member.IsAbstract || analysis.MakeDeclarationAtDestinationAbstract))
```

Comment:
i'm confused about the need for both these checks.  Why do you MakeDeclarationAtDestinationAbstract?  Is it not sufficient to just check if any member you're pulling up was abstract?